src/java/org/apache/cassandra/io/sstable/format/bti/BtiFormat.md

```diff
@@ -0,0 +1,991 @@
+the generic code to construct and read on-disk tries, and 
+`o.a.c.io.sstable.format.bti`, which implements the specifics of the
+format and the two indexes.
+
+## Building tries
+
+Tries are built from sorted keys using an [`IncrementalTrieWriter`](../../../tries/IncrementalTrieWriter.java). 
+The code contains three implementations with increasing complexity:
+- [`IncrementalTrieWriterSimple`](../../../tries/IncrementalTrieWriterSimple.java)
+  implements simple incremental construction of tries from sorted input,
+- [`IncrementalTrieWriterPageAware`](../../../tries/IncrementalTrieWriterPageAware.java)
+  adds packing of nodes to disk pages,
+- [`IncrementalDeepTrieWriterPageAware`](../../../tries/IncrementalDeepTrieWriterPageAware.java)
+  adds the ability to transition to on-heap recursion for all stages of the construction
+  process to be able to handle very large keys.
+
+Only the latter is used, but we provide (and test) the other two as a form of
+documentation.
+
+The builders take a `TrieSerializer` as parameter, which determines how the nodes
+are written. The indexes implement this using `TrieNode`, writing any payload they
+need immediately after the node serialization.
+
+## Reading tries
+
+The BTI format tries are used directly in their on-disk format. To achieve this,
+all node types are implemented as static objects in `TrieNode`. Reading nodes in
+a file is encapsulated in [`Walker`](../../../tries/Walker.java), 
+which provides a method to `go` to a specific node and use it, i.e. 
+get any associated data, search in the children list and
+follow transitions to children. It also provides functionality to find the
+mapping for a given key, floors and ceilings as well as some combinations.
+Iterating the payloads between two key bounds is implemented by 
+[`ValueIterator`](../../../tries/ValueIterator.java),
+and [`ReverseValueIterator`](../../../tries/ReverseValueIterator.java).
+
+Special care is given to prefixes to make sure the semantics of searches matches
+what the format needs.
+
+## SSTable format implementation
+
+The two indexes are implemented, respectively, by [`PartitionIndex`](PartitionIndex.java)
+and [`RowIndexReader`](RowIndexReader.java)/[`RowIndexWriter`](RowIndexWriter.java). 
+The format implementation extends the filtered
+base class and follows the structure of the BIG implementation, where
+all references to the primary index are replaced with calls to these two 
+classes.
+
+# Index file format in BTI
+
+## Trie nodes
+Implemented in [`TrieNode.java`](../../../tries/TrieNode.java)
+
+Nodes start with four bits of node type, followed by 4 payload bits
+(_pb_), which are 0 if the node has no associated payload; otherwise the
+node type gives an option to compute the starting position for the
+payload (_ppos_). The layout of the node depends on its type.
+
+`PAYLOAD_ONLY` nodes:
+
+-   4 type bits, 0
+
+-   4 payload bits
+
+-   payload if _pb_ &ne; 0, _ppos_ is _node_ + 1
```

Comment:
Perhaps we should use "position" or "node position" rather than just "node" here to indicate it's from the starting position of the node and this is calculated to exactly "skip over" the rest of the node contents.